### ISSUE TITLE
code missing

### ISSUE THREAD

**Issue by qiminchen** [paschalidoud/hierarchical_primitives#4]
Hi Despoina,

Thank you so much for updating the code. I noticed a few potential errors and missing in the code.

1. the name of folder `hierarchical_primitives` which has the same level as folder `scripts` i guess should be `learnable_primitives` https://github.com/paschalidoud/hierarchical_primitives/blob/e31524d3a7af7d947f91a9140210faaf114873d7/scripts/train_network.py#L26
2. some files are missing, for example, `sample_points_on_primitive` and `voxelizers` https://github.com/paschalidoud/hierarchical_primitives/blob/e31524d3a7af7d947f91a9140210faaf114873d7/scripts/train_network.py#L27 https://github.com/paschalidoud/hierarchical_primitives/blob/e31524d3a7af7d947f91a9140210faaf114873d7/scripts/train_network.py#L31

Would it be possible to have a readme about how to prepare our own dataset for training and how to run the script for training? We want to run some experiments and do comparisons. Any updates would be very appreciated.

**Comment by paschalidoud** [comment]
Hi @qiminchen,

I am really sorry for this!

I will try to update the README and also push the config files for both D-FAUST and the ShapeNet experiments from our paper throughout the day. I just committed a fix that should solve the errors you see.

Best,
Despoina

**Comment by qiminchen** [comment]
thanks Despoina, btw i think `hierarchical_primitives/sample_points_on_primitive.py` and `hierarchical_primitives/utils/pcl_voxelization.py` are missing, inferred from
https://github.com/paschalidoud/hierarchical_primitives/blob/852099e5fdd3625880567b359f63118e59bc2da2/scripts/train_network.py#L26 and https://github.com/paschalidoud/hierarchical_primitives/blob/852099e5fdd3625880567b359f63118e59bc2da2/hierarchical_primitives/voxelizers.py#L5
again, thank you so much for sharing the code, can't wait to try the work.

**Comment by paschalidoud** [comment]
I am sorry for this @qiminchen . I just committed the missing files. It should be fine now.

Best,
Despoina

**Comment by qiminchen** [comment]
Hi Despoina, please don't be sorry, I really appreciate your quick response. Would you please tell when will the ReadMe for training guidance be available? Will you release the training data? Thank you.

**Comment by qiminchen** [comment]
Hi Despoina,

I am wondering if you could update the config file for the experiment? Thanks

https://github.com/paschalidoud/hierarchical_primitives/blob/7c7b71ca1ffe998f8099bb6fed4a43fa32fba71c/scripts/train_network.py#L127-L131

**Comment by paschalidoud** [comment]
Hi @qiminchen,

I am really sorry for taking so long to reply. I just updated the config file and the README for the training script, so hopefully now you will be able to test the model. During the weekend, I will also upload the pre-trained models.

I am closing the issue for now, but please feel free to reopen it if you like!

Sorry again for the late reply!

Best,
Despoina

**Comment by qiminchen** [comment]
Hi Despoina,

Thanks for updating the repo, is it possible to provide the ShapeNet training data as well? I'm using my own preprocessed ShapeNet data (image, points, labels) to train the network from scratch but it gave me negative fit loss. Do you have any ideas why this happened? Thank you.

![Screenshot from 2021-03-29 15-07-02](https://user-images.githubusercontent.com/45860234/112906401-fb498a80-90a0-11eb-8172-d77b6be5183e.png)

**Comment by qiminchen** [comment]
Hi @paschalidoud,

Thanks for updating the repo, is it possible to provide the ShapeNet training data as well? I'm using my own preprocessed ShapeNet data (image, points, labels) to train the network from scratch but it gave me negative fit loss. Do you have any ideas why this happened? Thank you.

![Screenshot from 2021-03-29 15-07-02](https://user-images.githubusercontent.com/45860234/112906871-c7229980-90a1-11eb-9987-269cf9a5d6ba.png)

**Comment by paschalidoud** [comment]
HI @qiminchen,

Sorry for my late reply. Unfortunately I have never experienced negative fit loss. It actually doesn't make much sense.

The only difference between the shapenet config and the dfaust config file is that for shapenet the ``normalize`` key in the ``data`` field should be set to true. Furthermore, are your meshes watertight?

Best,
Despoina

**Comment by qiminchen** [comment]
Hi @paschalidoud 

Please dont be sorry, hmm we did not directly use mesh file (.obj), instead, we rendered points, label(occupancy value) and images first, and use them for training, this is probably problematic, is mesh (.obj) the only file needed for training? or do we need to render some other data, like images and points and occupancy value, before training?

**update**: I checked the model_factory.py and it seems like we need to render `_occupancy_pairs["points"], _occupancy_pairs["occupancies"], _surface_pairs["points"], _surface_pairs["occupancies"]` and corresponding images, it would be perfect if you could provide the script to render these files from mesh?

**Comment by paschalidoud** [comment]
Hi @qiminchen,

In  the [model_factory.py](https://github.com/paschalidoud/hierarchical_primitives/blob/66341ae9eb94d4f85be0f9ccc3f73c77332afd36/hierarchical_primitives/common/model_factory.py) file, the function that is used for implementing the sampling strategy "from" the target mesh is the[ sample_points()](https://github.com/paschalidoud/hierarchical_primitives/blob/66341ae9eb94d4f85be0f9ccc3f73c77332afd36/hierarchical_primitives/common/model_factory.py#L192) function. Based on the values of the arguments "equal", "positive weight" and "biased" that are specified in the config file you can select the type of sampling you want. To have equal unbiased sampling, which is the sampling that we used in the experiments of our paper, you simply need to set "equal: true".

Now, the function that is responsible for sampling the points in the bounding box (occupancy pairs) is the [get_points_in_bbox()](https://github.com/paschalidoud/hierarchical_primitives/blob/66341ae9eb94d4f85be0f9ccc3f73c77332afd36/hierarchical_primitives/common/model_factory.py#L108). To speed up training a bit this function does not compute the occupancy pairs but instead loads them from a file specified by the [path_to_occupancy_pair](https://github.com/paschalidoud/hierarchical_primitives/blob/66341ae9eb94d4f85be0f9ccc3f73c77332afd36/hierarchical_primitives/common/model_factory.py#L111) property and then keeps them in memory. Based on the dataset the path_to_occupancy_pair is different, as different datasets have different format. For the case of ShapeNet, the path_to_occupancy_pairs property should be a file with name points.npz, inside each folder for each model in the dataset.

For generating the occupancy pairs, you can use the script [generate_surface_samples.py](https://github.com/paschalidoud/neural_parts/blob/master/scripts/generate_surface_samples.py) with --occupancy type volume. I will also commit this script in the hierarchical_primitives repository so that you can directly use it from there! :-)

To summarize, you don't need to specify the obj file, only the path to occupancy pairs and the images directory. My speculation is that in your case the occupancy pairs are garbage from the disk thus you see negative loss.
